### DISCUSSION TITLE
Allow Passthrough Responses by Throwing a Request

### DISCUSSION THREAD

**Discussion by swamidass** [remix-run/remix#8305] [Proposals]
# Allow Passthrough Responses by Returning a Request

When building for the edge, we often want to pass through particular requests. For example, consider the case of static files collected in a particular route. If remix gets a request on this route, we'd want remix to avoid fetching the s3 all-together, and just return the request we want the cloudfront/cloudflare/fastly distribution to use.

That's the easy case for which there are some platform specific workarounds. But more complex cases might have Remix applying more complicated logic to alter the request. 

This seems to be a key capability given remix's goal to be built for the Edge. 

## The problem

- Currently, there isn't an easy way of instructing the edge distribution to request from another origin.
- All the ways of doing this are are platform specific.
- There isn't a standardized approach, so plugin authors aren't covering this case.
- None of the work arounds work for local dev.

## Proposed solution

- Allow remix loaders to return a standard request object.
- In the dev server, request responses would be proxied through and fetched directly.
- In plugins for particular platforms, this request would be marshaled into the platform specific format and API for an origin request response.

## No Disruption to the current API

As far as I can tell it would leave everything else intact, without any disruption. Any projects that don't use this feature would not be impacted. Currently request responses are not used for anything else.

## Not Difficult to Implement?

This should be straight forward to implement as it's logic can be confined entirely within a single special case in (1) the adapter plugins and (2) the dev server. That means would be releasable without requiring a feature-flag for backward compatibility. 

## Other options?

There may be a better way to accomplish this. If so, please let me know. Feel free to edit or revise this proposals as needed.

-----

NOTE: Perhaps this should be throwing a request instead of returning one.

**Comment by kiliman** [comment]
I'm not quite sure what issue you're having. It may be helpful to provide some examples. Remix loaders and actions return standard `Response` objects. If you want to pass through a request upstream, you can simply return a call to `fetch`.

```ts
export async function loader() {
  return fetch(...)
}
```

**Reply by swamidass** [reply]
Thanks for the response!

The problem with the suggested approach (of returning the results of a fetch) is that it keeps the edge function running for the duration of the fetch. This has major drawbacks:

- This increases the costs, in which edge run-time pricing is steep.
- lambda runs is subject to a 5 second total time, so low latency or long polling sorts of origins don't work.
- It entirely ignores the retry policies and infrastructure of the platform 

In contrast, returning a request back to the platform immediately terminates the function. This has some really important advantages.

- function runtime is keep to a minimum, and the origin request time is entirely free.
- far longer time outs are possible (e.g. 60 seconds in origin) 
- the retry policies can be configured for the origin
- the infrastructure of the platform in making origin requests is leveraged for substantial performance gains.
